20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,171 @@
+
+### Syntax
+
+Let A be an authorization model with set of types T and set of relations R, and recall that R* is the set R with the addition of an "empty" relation. We will write an element (t, r) of TxR* as t#r. The special case (t, _) is simply written as t. If the type restriction takes (t, r) to a set {(t1, r1), (t2, r2)}, then in our model we will write this as
+```
+type t
+  relations
+    define r: [t1#r1, t2#r2] ...
+```
+If the type restriction takes (t, r) to the empty set we will simply write nothing in our model.
+
+### Types with no relations
+
+It now becomes useful to allow types with no relations. For example, suppose we want to define a `viewer` relation and retrict the typeswe may want to specify a `user` type which is the type for all concrete users. This will look like
+```
+type user
+```
+
+### The dropping of "as" and "self"
+
+In the new DSL we will drop "as" and "self". We drop "as" by adding ":" to all relations, even ones with no type restrictions. That is, if previously we had `define editor as owner`, this becomes `define editor: owner`. With the addition of type restrictions means we may also drop `self` from the DSL. This is because if a pair (type, relation) has a non-empty set of type restrictions, then the assumption is that (type, relation) is a pair that allows a direct relation to type. Thus, previously if we defined a relation as
+```
+define viewer as self
+```
+we can now write this as
+```
+define viewer: [user]
+```
+
+The dropping of "self" will be extended to rewrite rules as well. If a rewrite contains `self` then it may be dropped as in the following examples:
+- `define viewer as self or viewer from parent` -> `define viewer: [user] or viewer from parent`
+- `define viewer as self and viewer from parent` -> `define viewer: [user] and viewer from parent`
+- `define viewer as self but not banned` -> `define viewer: [user] but not banned`
+
+### Wildcards
+
+The syntax and semantics of wildcards have also been updated in the new DSL. Consider the (new) authorization model
+```
+type user
+
+type document
+  relations
+    define viewer: [user]
+```
+Previously, writing a tuple with the wildcard symbol `document:X#viewer@*` meant that we intended the object `document:X` to be publically "viewable". In the new DSL, we will also add type restrictions to the wildcard symbol, and there will no longer be an untyped wildcard symbol. If you want to say that a document can be publically viewable for all `user`s, then you must add `user:*` to the type restrictions:
+```
+type user
+
+type document
+  relations
+    define viewer: [user, user:*]
+```
+Then you can write tuples such as `document:X#viewer@user:*`, which means that `document:X` is viewable to all `user`s. If you had another type, say `employee`, then you may instead write something like
```

Comment:
If you had another type __restriction__, say `employee`

Reply:
Thanks!